src/providers/wfs/qgsbasenetworkrequest.cpp

```diff
@@ -0,0 +1,466 @@
+  , mForceRefresh( false )
+  , mTimedout( false )
+  , mGotNonEmptyResponse( false )
+{
+  connect( QgsNetworkAccessManager::instance(), qgis::overload< QNetworkReply *>::of( &QgsNetworkAccessManager::requestTimedOut ), this, &QgsBaseNetworkRequest::requestTimedOut );
+}
+
+QgsBaseNetworkRequest::~QgsBaseNetworkRequest()
+{
+  abort();
+}
+
+void QgsBaseNetworkRequest::requestTimedOut( QNetworkReply *reply )
+{
+  if ( reply == mReply )
+    mTimedout = true;
+}
+
+bool QgsBaseNetworkRequest::sendGET( const QUrl &url, const QString &acceptHeader, bool synchronous, bool forceRefresh, bool cache )
+{
+  abort(); // cancel previous
+  mIsAborted = false;
+  mTimedout = false;
+  mGotNonEmptyResponse = false;
+
+  mErrorMessage.clear();
+  mErrorCode = QgsBaseNetworkRequest::NoError;
+  mForceRefresh = forceRefresh;
+  mResponse.clear();
+
+  QUrl modifiedUrl( url );
+
+  // Specific code for testing
+  if ( modifiedUrl.toString().contains( QLatin1String( "fake_qgis_http_endpoint" ) ) )
+  {
+    // Just for testing with local files instead of http:// resources
+    QString modifiedUrlString = modifiedUrl.toString();
+    // Qt5 does URL encoding from some reason (of the FILTER parameter for example)
+    modifiedUrlString = QUrl::fromPercentEncoding( modifiedUrlString.toUtf8() );
+    QgsDebugMsgLevel( QStringLiteral( "Get %1" ).arg( modifiedUrlString ), 4 );
+    modifiedUrlString = modifiedUrlString.mid( QStringLiteral( "http://" ).size() );
+#ifdef Q_OS_WIN
+    // Passing "urls" like "http://c:/path" to QUrl 'eats' the : after c,
+    // so we must restore it
+    if ( modifiedUrlString[1] == '/' )
+    {
+      modifiedUrlString = modifiedUrlString[0] + ":/" + modifiedUrlString.mid( 2 );
+    }
+#endif
+
+    // For REST API using URL subpaths, normalize the subpaths
+    int afterEndpointStartPos = modifiedUrlString.indexOf( "fake_qgis_http_endpoint" ) + strlen( "fake_qgis_http_endpoint" );
+    QString afterEndpointStart = modifiedUrlString.mid( afterEndpointStartPos );
+    afterEndpointStart.replace( QLatin1String( "/" ), QLatin1String( "_" ) );
+    modifiedUrlString = modifiedUrlString.mid( 0, afterEndpointStartPos ) + afterEndpointStart;
+
+    if ( !acceptHeader.isEmpty() )
+    {
+      if ( modifiedUrlString.indexOf( '?' ) > 0 )
+      {
+        modifiedUrlString += QStringLiteral( "&Accept=" ) + acceptHeader;
+      }
+      else
+      {
+        modifiedUrlString += QStringLiteral( "?Accept=" ) + acceptHeader;
+      }
+    }
+    auto posQuotationMark = modifiedUrlString.indexOf( '?' );
+    if ( posQuotationMark > 0 )
+    {
+      QString args = modifiedUrlString.mid( posQuotationMark );
+      if ( modifiedUrlString.size() > 256 )
+      {
+        QgsDebugMsgLevel( QStringLiteral( "Args before MD5: %1" ).arg( args ), 4 );
+        args = QCryptographicHash::hash( args.toUtf8(), QCryptographicHash::Md5 ).toHex();
+      }
+      else
+      {
+        args.replace( QLatin1String( "?" ), QLatin1String( "_" ) );
```

Comment:
could all be replaced by character based methods for speed

Reply:
taken into account